data/moves.ts

```diff
@@ -10108,10 +10110,18 @@ export const Moves: {[moveid: string]: MoveData} = {
 		priority: 0,
 		flags: {snatch: 1, gravity: 1},
 		volatileStatus: 'magnetrise',
+		onTry(source, target, move) {
+			if (target.volatiles['smackdown'] || target.volatiles['ingrain']) return false;
+
+			// Additional Gravity check for Z-move variant
```

Comment:
What I suspect you need is an additional event handler (which event depends on when the Z-move fails relative to other effects such as Protean) for Gravity to use to block these moves. It does depend somewhat on whether Gravity should be blocking called airborne moves. (For instance, you could hack a Pokémon with Protean, Sleep Talk and Telekinesis while its ally sends it to sleep and sets Gravity. Similarly, should Throat Chop block called sound moves or Heal Block block called healing moves, and if so, at what point do they fail, relative to other effects such as Protean?)

Reply:
Like I linked in my pastebin above, the game literally does these Gravity checks in the same event handler as the others effects I touched, so I think it's more accurate to include them here directly.

Gravity, Throat Chop, Heal Block, and whether a move is blocked in a Sky Battle are all ran on submoves. They fail in the pre-move success checks The move proper would deduct PP, but the user won't become Choice locked or switch from Stance Change or set a target. Some tests from my Google Doc (this is Gen 8)

first action is Sleep Talk while asleep -> calls Snore -> blocked by Throat Chop -> Pokémon not locked into Sleep Talk (PP deducted)
Sleep Talk while asleep -> calls Snore -> blocked by Throat Chop -> (1 PP deducted despite both opponents having Pressure)
Sleep Talk while asleep -> calls Snore into Pressure -> 2 PP deducted

SadisticMystic's SM notes say Choice locking occurred prior to submove calls in 7.

Reply:
It's the triplication of code that I was most concerned about, but the centralisation would also make it easy to adapt for called airborne moves under Gravity.

As for ensuring that the Gravity event handler runs at the appropriate time, based on my reading of your comments and links, this actually needs to be `onModifyMove` anyway so that it happens before the move is announced and targetted. (Certainly for Telekinesis the Smack Down and Ingrain check can't happen until after the move is targetted!)

(Note that currently on the sim Stance Change never triggers for Sleep Talk calling an attack, so if that's wrong it probably also needs to switch to using the `onModifyMove` event.)

Reply:
> It's the triplication of code that I was most concerned about, but the centralisation would also make it easy to adapt for called airborne moves under Gravity.
That's a fair point, and one you could have poked Game Freak about. :P

> As for ensuring that the Gravity event handler runs at the appropriate time, based on my reading of your comments and links, this actually needs to be `onModifyMove` anyway so that it happens before the move is announced and targetted. (Certainly for Telekinesis the Smack Down and Ingrain check can't happen until after the move is targetted!)
Strange as it might be, the Gravity check really is supposed to be here: it really does announce the move, set the target, and deduct PP. Here's a video of behavior with me targeting a Pressure Pokemon and losing 2 PP while Z-Telekinesis gets stopped by Gravity: https://www.youtube.com/watch?v=Lppqg2ifVR8 The Gravity checks really are just are in this location.

> (Note that currently on the sim Stance Change never triggers for Sleep Talk calling an attack, so if that's wrong it probably also needs to switch to using the `onModifyMove` event.)
TI fixed this in an independent commit!

Reply:
> That's a fair point, and one you could have poked Game Freak about. :P

So they only special-cased those three moves? What about Supersonic Skystrike? Or does some other check defeat it? What if you try to (say) Z-Assist a Supersonic Skystrike?

> https://www.youtube.com/watch?v=Lppqg2ifVR8

Sorry, I misunderstood that the two documents needed to be read in combination.

Reply:
Even after rewriting several times that still came out wrong. I meant I misunderstood how the documents needed to be combined.

Reply:
All good, it's very bizarre (and newly discovered!) behavior.

> So they only special-cased those three moves? What about Supersonic Skystrike? Or does some other check defeat it? What if you try to (say) Z-Assist a Supersonic Skystrike?

That's right; Z-Fly, Z-Bounce etc. wouldn't be impacted by this. It's only an explicit Gravity check for those 3 moves.

Reply:
(What was also confusing me is that Supersonic Skystrike bypasses Gravity, but it does that anyway because it's not marked as airborne!)